src/s2s/mongoose_s2s_out.erl

```diff
@@ -0,0 +1,593 @@
+            handle_socket_packet(Data, Packet)
+    end.
+
+-spec handle_socket_packet(data(), binary()) -> fsm_res().
+handle_socket_packet(#s2s_data{parser = Parser} = Data, Packet) ->
+    case exml_stream:parse(Parser, Packet) of
+        {error, Reason} ->
+            NextEvent = {next_event, internal, #xmlstreamerror{name = Reason}},
+            {keep_state, Data, NextEvent};
+        {ok, NewParser, XmlElements} ->
+            Size = iolist_size(Packet),
+            NewData = Data#s2s_data{parser = NewParser},
+            handle_socket_elements(NewData, XmlElements, Size)
+    end.
+
+-spec handle_socket_elements(data(), [exml_stream:element()], non_neg_integer()) -> fsm_res().
+handle_socket_elements(#s2s_data{myname = LServer, shaper = Shaper} = Data, Elements, Size) ->
+    {NewShaper, Pause} = mongoose_shaper:update(Shaper, Size),
+    [mongoose_instrument:execute(
+       xmpp_element_size_in, labels(),
+       #{byte_size => exml:xml_size(Elem), lserver => LServer, pid => self(), module => ?MODULE})
+     || Elem <- Elements],
+    NewData = Data#s2s_data{shaper = NewShaper},
+    StreamEvents0 = [ {next_event, internal, XmlEl} || XmlEl <- Elements ],
+    StreamEvents1 = maybe_add_pause(NewData, StreamEvents0, Pause),
+    {keep_state, NewData, StreamEvents1}.
+
+-spec maybe_add_pause(data(), [gen_statem:action()], integer()) -> [gen_statem:action()].
+maybe_add_pause(_, StreamEvents, Pause) when Pause > 0 ->
+    [{{timeout, activate_socket}, Pause, activate_socket} | StreamEvents];
+maybe_add_pause(#s2s_data{socket = Socket}, StreamEvents, _) ->
+    mongoose_xmpp_socket:activate(Socket),
+    StreamEvents.
+
+%% This will trigger if after a while the gen_statem state has not changed.
+%% Should be enabled for all transitions except into `stream_established`.
+-spec state_timeout(data() | options()) -> {state_timeout, timeout(), state_timeout_termination}.
+state_timeout(#s2s_data{opts = Opts}) ->
+    state_timeout(Opts);
+state_timeout(#{state_timeout := Timeout}) ->
+    {state_timeout, Timeout, state_timeout_termination}.
+
+%% This will trigger _only_ during `stream_established`, to garbage-collect unused connections.
+-spec stream_timeout(data() | options()) -> {timeout, timeout(), stream_timeout}.
+stream_timeout(#s2s_data{opts = Opts}) ->
+    stream_timeout(Opts);
+stream_timeout(#{stream_timeout := Timeout}) ->
+    {timeout, Timeout, stream_timeout}.
+
+%% Sockets failed to connect entirely
+%% The initial delay is random between 1 and `max_retry_delay` seconds
+-spec wait_before_retry_timeout(data() | options()) ->
+    {{timeout, wait_before_retry}, timeout(), wait_before_retry_timeout}.
+wait_before_retry_timeout(#s2s_data{opts = Opts}) ->
+    wait_before_retry_timeout(Opts);
+wait_before_retry_timeout(#{max_retry_delay := MaxRetryDelay}) ->
+    Delay = 999 + rand:uniform(timer:seconds(MaxRetryDelay) - 999),
+    {{timeout, wait_before_retry}, min(Delay, MaxRetryDelay), wait_before_retry_timeout}.
```

Comment:
min seems to be unnecessary here. `Delay` is always smaller than or equal to `MaxRetryDelay`.

Reply:
and I belive MaxRetryDelay should be converted into milliseconds before comparison.